rl_algorithms/recurrent/dqn_agent.py

```diff
@@ -0,0 +1,265 @@
+                self.memory, alpha=self.hyper_params.per_alpha
+            )
+
+            # replay memory for multi-steps
+            if self.use_n_step:
+                self.memory_n = RecurrentReplayBuffer(
+                    self.hyper_params.buffer_size,
+                    self.hyper_params.batch_size,
+                    self.hyper_params.sequence_size,
+                    self.hyper_params.overlap_size,
+                    n_step=self.hyper_params.n_step,
+                    gamma=self.hyper_params.gamma,
+                )
+
+        self.learner = build_learner(self.learner_cfg)
+
+    def select_action(
+        self,
+        state: np.ndarray,
+        hidden_state: torch.Tensor,
+        prev_action: torch.Tensor,
+        prev_reward: np.ndarray,
+    ) -> np.ndarray:
+        """Select an action from the input space."""
+        self.curr_state = state
+
+        # epsilon greedy policy
+        # pylint: disable=comparison-with-callable
+        state = self._preprocess_state(state)
+        selected_action, hidden_state = self.learner.dqn(
+            state, hidden_state, prev_action, prev_reward
+        )
+        selected_action = selected_action.detach().argmax().cpu().numpy()
+        if not self.args.test and self.epsilon > np.random.random():
+            selected_action = np.array(self.env.action_space.sample())
+        return selected_action, hidden_state
+
+    def _add_transition_to_memory(self, transition: Tuple[np.ndarray, ...]):
+        """Add 1 step and n step transitions to memory."""
+        # add n-step transition
+        if self.use_n_step:
+            transition = self.memory_n.add(transition)
+
+        # add a single step transition
+        # if transition is not an empty tuple
+        if transition:
+            self.memory.add(transition)
+
+    def step(
+        self, action: np.ndarray, hidden_state: torch.Tensor
+    ) -> Tuple[np.ndarray, np.float64, bool, dict]:
+        """Take an action and return the response of the env."""
+        next_state, reward, done, info = self.env.step(action)
+        if not self.args.test:
+            # if the last state is not a terminal state, store done as false
+            done_bool = (
+                False if self.episode_step == self.args.max_episode_steps else done
+            )
+
+            transition = (
+                self.curr_state,
+                action,
+                hidden_state.detach(),
+                reward,
+                next_state,
+                done_bool,
+            )
+            self._add_transition_to_memory(transition)
+
+        return next_state, reward, done, info
+
+    def sample_experience(self) -> Tuple[torch.Tensor, ...]:
+        experience_1 = self.memory.sample(self.per_beta)
+        if self.use_n_step:
+            indices = experience_1[-2]
+            experience_n = self.memory_n.sample(indices)
+            return numpy2floattensor(experience_1), numpy2floattensor(experience_n)
```

Comment:
If memory is PER, experience has indices and eps_d. They don't have to be changed numpy to tensor.
It needs to be fixed.